### ISSUE TITLE
Update GP chapter in Stan User Guide

### ISSUE THREAD

**Issue by mike-lawrence** [stan-dev/docs#614]
Chapter 10 of the Stan User Guide currently uses either custom code or the deprecated `cov_exp_quad()` where `gp_exp_quad_cov()` should be used. Also, might be worth showing the use of the the convenience functions from Aki's latest paper [here](https://github.com/gabriuma/basis_functions_approach_to_GP/blob/a5e6a1aa82eb206fa4d235c0f9844d17ef29c337/Paper/Case-study_Same-sex-marriage-data/CaseV_Same-sex-marriage_Exact-GP.stan#L80) and [here](https://github.com/gabriuma/basis_functions_approach_to_GP/blob/4199949163e71a6f3b8f51ea5f7e9fee6a2d9fc9/Paper/Case-study_1D-Simulated-data/first_version/CaseI_1D-Simulated-data_Exact-GP.stan#L47)

**Comment by bob-carpenter** [comment]
Thanks, @mike-lawrence.  That sounds good.

What we really need to do is move to comprehensions with partially evaluated GP autodiff!  To the extent that we have to use primitive covariance matrix constructing functions and add them, we wind up with a huge expression graph for autodiff that limits scalability. We could gin something up now with roughly the right functionality using one of the parallel `map` functions, because they do partial evaluation.  It'd just be a fussy indexing mess.

**Comment by bob-carpenter** [comment]
> This bit pertains to enhancing Stan, and not the documentation thereof, correct?

Both.  It'd be nice if we had a matrix comprehension function, but in the meantime, one can be implemented using `map_rect`.  The advantage is that we never build a full covariance matrix of expressions, but instead evaluate each one's derivative on the fly.  The parallelism is just gravy, but super helpful if you're running 4 chains on a 32 core server or even 2 chains on a 4 core server.

That is, we'd have something like `map_rect` where the function being mapped is the covariance function.  It has to compute entry `n` in the covariance matrix.  But it returns a vector, which then needs to be reshaped into a matrix for use.  That's what I meant by being fiddly in terms of indexing.  For an M x N matrix indexed from 1, entry (m, n) is in position 

```
position(m, n, M, N) = (M * (m - 1)) + (n - 1) + 1.  
```

So the inverse is fiddly.  

Where this will make a big difference is in something like the "birthday problem" from the cover of *BDA3*, the covariance matrix for which is the sum of a sequence of other covariance matrix.  So each entry has a large-ish expression graph.  It'll also make the difference between doable and not doable when you're near memory limits.

**Comment by bob-carpenter** [comment]
We are going to be deprecating the `#` comments, so I'd suggest switching to `//` line comments.  

The nice thing about `cov_exp_quad()` is that we can write analytic gradients for the whole function.  The nice thing about mapping is that it saves a lot of memory and improves memory locality.  The ideal combination would have the special functions be functions that produce entries with analytic derivatives.  

Anyway, this is probably too much for a user's guide chapter, but something that'd make a nice case study.

**Comment by mike-lawrence** [comment]
> We are going to be deprecating the `#` comments, so I'd suggest switching to `//` line comments.

Yup, for sure; that gist is 6 years old; a bunch of other updates to be made there too.
